### PR TITLE
3296 location mvp mvp

### PR DESCRIPTION
**Pull request by easherma**
# Description

By popular demand! A naive implementation of location page, with lots of copy pasta, and a copy pasta factory to make you more pasta.

# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How can this be tested?

 

# Checklist:
- [ ] Request reviewers
- [ ] Slack-out message for review
- [ ] Link this PR in the issue
- [ ] Moved card to *review* in zenhub

### CODE REVIEW

File: joplin/base/templatetags/joplin_tags.py [hunk 1]
```diff
@@ -32,7 +34,11 @@ def get_revision_preview_url(*args, **kwargs):
 @register.simple_tag
 def get_style_guide_url(*args, **kwargs):
     content_type = kwargs['content_type'].name
-    return os.environ['STYLEGUIDE_URL'] + STYLEGUIDE_PAGES[content_type]
+    try:
+        get_style_guide_url = os.environ['STYLEGUIDE_URL'] + STYLEGUIDE_PAGES[content_type]
+    except Exception as e:
```
Comment (easherma) [inline joplin/base/templatetags/joplin_tags.py]: gives a default fallback for a page if its missing a styleguide url

File: joplin/base/models/janis_page.py [hunk 2]
```diff
@@ -165,11 +164,14 @@ def janis_preview_url_end(self, revision=None):
         """
         if revision:
             url_page_type = revision.page.janis_url_page_type
+            global_id = graphene.Node.to_global_id('PageRevisionNode', revision.id)
```
Comment (easherma) [inline joplin/base/models/janis_page.py]: @Niicck ran into some errors when making LocationPage a JanisPage (where I think there werent any revisions yet?)
anyways, this seems to do what you were looking to do while avoiding triggering an error if global_id fails for not existing

File: joplin/base/wagtail_hooks.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def configure_main_menu(request, menu_items):
     menu_items[:] = [item for item in menu_items if item.name not in
                      # here were excluding some default generated menu items per UX
                      [
-                         'explorer',
+                         # 'explorer',
```
Comment (easherma) [inline joplin/base/wagtail_hooks.py]: we can re-hide this later, atm I'm using it as an interface to show how index pages might work. in another branch i have an example using ModelAdmin, but thats probably overkill for this situation

File: joplin/base/models/janis_page.py (hunk 2 above)
Comment (Niicck) [reply joplin/base/models/janis_page.py]: Ah, nope different logic going on.
```
Optional "revision" parameter to get the janis_preview_url for a specific revision
Otherwise, it will return the janis_preview_url for the latest revision
```
Now you've changed it so that the default preview for a page won't preview the latest revision.

This logic will have to be changed back. If there are errors getting a preview_url for LocationPage that doesn't have a revision, it needs to be handled in a different way.

File: joplin/base/templatetags/joplin_tags.py (hunk 1 above)
Comment (Niicck) [reply joplin/base/templatetags/joplin_tags.py]: You mean if its missing a STYLEGUIDE_PAGES[content_type] ?

File: joplin/locations/fixtures.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        this is probably a bad way to do this, we don't want to accidentially delete
+        data we care about, right?
+        """
+        try:
+            import locations
+            locations.models.LocationPage.objects.all().delete()
+        except Exception as e:
+            print(e)
+            import pdb
+            pdb.set_trace()
```
Comment (Niicck) [inline joplin/locations/fixtures.py]: We want to take out pdb.set_trace() in the Exception, even for mvp-mvp.

Is there more specific cleanup/handling to be done if an exception is caught here?

And why is this probably a bad way to handle clean up?

Comment (easherma) [reply joplin/locations/fixtures.py]: I think its 'probably' bad becuase its buried in a script that might get called in different contexts but it deletes all pages of that type in the database

Sooo say you have data on your branch, and then want to call this script to add test data. It would make test data on top of your data, than delete ALL data for the page type. That seems no bueno to me.

File: joplin/base/models/janis_page.py (hunk 2 above)
Comment (easherma) [reply joplin/base/models/janis_page.py]: Ok, my bad. Good to know. I'll revisit this.

File: joplin/base/templatetags/joplin_tags.py (hunk 1 above)
Comment (easherma) [reply joplin/base/templatetags/joplin_tags.py]: Yes

File: joplin/base/models/janis_page.py (hunk 2 above)
Comment (easherma) [reply joplin/base/models/janis_page.py]: @Niicck this should be correct now

File: joplin/base/models/janis_page.py [hunk 5]
```diff
@@ -257,6 +260,9 @@ def get_edit_handler(cls):
     class Meta:
         abstract = True
 
+# TODO: tech debt here, added an additional rule to hide this for location page
```
Comment (chiaberry) [inline joplin/base/models/janis_page.py]: Where did you add the additional rule?

also my branch (that I just merged to master) has some changes here, maybe that can be extended to hide for location

File: joplin/js/editor.js [hunk 6]
```diff
@@ -387,4 +396,36 @@ $(function() {
...
+
+  // autocomplete for places
+  (function() {
+    var latlng = {
+      lat: 30.26,
+      lng: -97.73,
+    };
+    var placesAutocomplete = places({
+      appId: 'plHX3G9C5GRN',
+      apiKey: 'd9275f5f90784eb5ba59ff8ec581d9bb',
```
Comment (chiaberry) [inline joplin/js/editor.js]: should we hide these keys?

File: joplin/js/editor.js [hunk 7]
```diff
@@ -387,4 +396,36 @@ $(function() {
...
+      container: document.querySelector('#id_physical_street'),
+      templates: {
+        value: function(suggestion) {
+          return suggestion.name;
+        },
+      },
+    }).configure({
+      aroundLatLng: latlng.lat + ',' + latlng.lng, // Paris latitude longitude
```
Comment (chiaberry) [inline joplin/js/editor.js]: paris? its a leftover comment, right?

File: joplin/js/editor.js (hunk 6 above)
Comment (easherma) [reply joplin/js/editor.js]: we could, or import them as an env variable. but a)it was something slapped together as a quick POC, I haven't confirmed that we even want to use this particular service and b) im pretty sure the key would be exposed client side either way, i dont think its a real high priority at the moment but when we refactor we could address it

File: joplin/js/editor.js (hunk 7 above)
Comment (easherma) [reply joplin/js/editor.js]: Oui, good catch!

File: Pipfile [hunk 8]
```diff
@@ -30,6 +30,7 @@ phonenumberslite = "==8.10.16"
 django-debug-toolbar = "==2.0"
 django-countable-field = {editable = true,git = "https://github.com/cityofaustin/django-countable-field"}
 django-flags = "*"
+wagtail-factories = "*"
```
Comment (Niicck) [inline Pipfile]: Can you add all the dependencies you're going to need to the Pipfile as well? I've got a change coming that's going to use Pipfile as the single source of truth. And maybe things like Faker can go into dev-dependencies.

File: joplin/base/models/janis_page.py [hunk 9]
```diff
@@ -168,8 +167,12 @@ def janis_preview_url_end(self, revision=None):
         else:
             revision = self.get_latest_revision()
             url_page_type = self.janis_url_page_type
-        global_id = graphene.Node.to_global_id('PageRevisionNode', revision.id)
-        url_end = f"preview/{url_page_type}/{global_id}"
+        try:
+            global_id = graphene.Node.to_global_id('PageRevisionNode', revision.id)
+            url_end = f"preview/{url_page_type}/{global_id}"
+        except AttributeError:
+            url_end = f"preview/{url_page_type}/"
```
Comment (Niicck) [inline joplin/base/models/janis_page.py]: Sorry I forgot, when does this happen again? If a page doesn't have a revision yet? Do locations not have revisions?

If this exception has been added just for the sake of MVP locations, let's make a "TODO" note to clean up previews on locations and then get rid of this logic. Because a preview of a page that has no revision should never exist.

File: Pipfile (hunk 8 above)
Comment (easherma) [reply Pipfile]: this is the only new dependency I added, it uses Faker as a dependency directly 

All the changes in requirements.txt are becuase I re-froze it, and pip sorts requirements.txt alphabetically by default 

FYI @briaguya this addresses your question about requirements.txt as well

File: joplin/base/models/janis_page.py (hunk 9 above)
Comment (easherma) [reply joplin/base/models/janis_page.py]: IIRC it had to do with loading fixtures or generating test pages, those pages exist but I think they didn't have revisions associated with them yet. So it could happen with any page that is created but not with an associated revision. 

Example test page:
https://joplin-pr-3296-location-mvp-mv.herokuapp.com/admin/pages/267/revisions/

A location page has revisions like any other Wagtail Page type. 

Since we're writing our own url logic here, I think it's our responsibility to handle edge cases. In this case I wouldn't expect this to ever be a problem in staging or prod unless we progamatically enter content at some point(which could very well happen), but the try statement here has no performance drawbacks but helps avoid unexpected problems that might break the admin. 

From my perspective, good candidates for refactoring in this case would be how we handle previews (using https://github.com/torchbox/wagtail-headless-preview would be nice) and re-tooling to leverage wagtails existing functions for building urls which are more robust (or least informing our design based on them: 
https://github.com/wagtail/wagtail/blob/8238a338888258554ed6eae0e25eaf903a0a5831/wagtail/core/models.py#L748 )

Comment (Niicck) [reply joplin/base/models/janis_page.py]: But this doesn't handle the edge case. It's just going to allow previews to break for test fixtures.

If fixtures don't have revision data, that should be accounted for in logic farther up the line rather than returning a preview_url that won't work.

But for the sake of getting the locations MVP merged in, let's worry about that later.

Comment (easherma) [reply joplin/base/models/janis_page.py]: Yes, to be clear, it doesn't make previews work(which I didn't really have in scope for MVP as far as I'm aware), but without this try statement the entire build was breaking.
